features/256-auth-gear-interaction-overview-in-next/api.md

```diff
@@ -0,0 +1,441 @@
+
+| Attribute | Interface | Old JS SDK API |
+| -------- | -------- | ------ |
+| disabled | `POST /auth/disable/set`| `adminDisableUser`<br/>`adminEnableUser` |
+| password | `POST /auth/reset_password`| `adminResetPassword` |
+| role | `POST /auth/role/assign`<br/>`POST /auth/role/admin`<br/>`POST /auth/role/default` | `assignUserRole`<br/>`revokeUserRole`<br/>`setDefaultRole` |
+
+For security reason, Cloud Function should invoke admin related API with master key.
+
+## Execution flow of hooked Cloud Function
+
+There are four hooked Cloud Function forms: 
+
+- `before_XXX_sync`
+- `before_XXX`
+- `after_XXX_sync`
+- `after_XXX`
+
+And as the name inferred, hooked Cloud Functions are executed in two ways: `sync` and `async` way (without `sync` suffix implies `async`). All hooks are executed in transaction. `before_XXX_sync` and `before_XXX` is executed "before" DB operation. Whereas `after_XXX_sync` and `after_XXX` is executed "after" DB operation. `before_XXX_sync` and `after_XXX_sync` can raise exception to abort current operation.
+
+Following pseudo code demonstrates the execution flow in auth gear:
+
+```go=
+// start DB transaction
+txContext.BeginTx()
+
+resp, err := handle(payload, &user)
+if err != nil {
+    response.Err = skyerr.MakeError(err)
+    h.TxContext.RollbackTx()
+    return response
+}
+
+err = hooks.ExecuteBeforeSyncHooks(&user)
+if err != nil {
+    response.Err = skyerr.MakeError(err)
+    h.TxContext.RollbackTx()
+    return response
+}
+
+hooks.ExecuteBeforeAsyncHooks(user)
+
+// DB operation
+err = userStore.update(user)
+if err != nil {
+    response.Err = skyerr.MakeError(err)
+    h.TxContext.RollbackTx()
+    return response
+}
+
+hooks.ExecuteAfterAsyncHooks(user)
+
+err = hooks.ExecuteAfterSyncHooks(user)
+if err != nil {
+    response.Err = skyerr.MakeError(err)
+    h.TxContext.RollbackTx()
+    return response
+}
+
+// DB commit
+txContext.CommitTx();
+
+response.Result = resp
+return response
+```
+
+In `before_XXX_sync` and `before_XXX`, it may alter user metadata, on the contrary, `after_XXX_sync` and `after_XXX` won't support to alter user metadata. And all hooks won't support alter user auth data.
+
+|  | alter metadata | alter auth data | raising an exception to stop operation |
+| -------- | -------- | -------- | ------ |
+| `before_XXX_sync`  | ✓ | 🚫 | ✓ |
+| `before_XXX`  | 🚫 | 🚫 | 🚫 |
+| `after_XXX_sync`  | 🚫 | 🚫 | ✓ |
+| `after_XXX`  | 🚫 | 🚫 | 🚫 |
+
+
+Function signature of `before_XXX_sync` (`after_XXX_sync` is similar) hooked Cloud Function is:
+
+```javascript
+const skygear = require('skygear');
+
+/* 
+ * user: user object to be saved
+ * context: current exection context
+ */
+function before_XXX_sync(user, context) {
+    console.log(user.metadata.loveCat); // false
+    
+    // alter user metadata
+    user.metadata.loveCat = true;
+    
+    /*
+     * or rasie exception
+     * throw new Error("some error");
```

Comment:
Instead of js `Error` object, I think we can use `SkygearError` here to provide error code too?